main.go

```diff
@@ -182,12 +185,64 @@ func wrappedMain() int {
 		}
 	}
 
+	unmanagedProvidersStr := os.Getenv("TF_REATTACH_PROVIDERS")
+	unmanagedProviders := map[addrs.Provider]*plugin.ReattachConfig{}
+	if unmanagedProvidersStr != "" {
+		// {"hashicorp/random":{"Protocol":"grpc","Addr":{"Name":"/tmp/plugin553215095","Net":"unix"},"Pid":19594,"Test":true}}
+		type reattachConfig struct {
+			Protocol string
+			Addr     struct {
+				Network string
+				String  string
+			}
+			Pid  int
+			Test bool
+		}
+		var m map[string]reattachConfig
+		err := json.Unmarshal([]byte(unmanagedProvidersStr), &m)
+		if err != nil {
+			Ui.Error("Invalid format for TF_REATTACH_PROVIDERS:" + err.Error())
```

Comment:
For environment variable problems we tend to log them and otherwise silently ignore them, because environment variables are often set ambiently in the shell environment and it's annoying to have Terraform be blocked by them being there. With that said, I don't think we can safely log in this context (I think the logger isn't configured properly yet here?) and I suppose this particular environment variable is unlikely to be set by anything other than the test framework, so maybe this is fine. :thinking: 

If we could, it might be nice to have this codepath always "succeed", but then raise an error when instantiating the provider in the `command` package later, so that we can do error handing for provider instantiation at a more conventional place. However, I can see that for this _particular_ error -- that the environment variable itself is just entirely invalid -- we'd end up having to just fail _all_ provider instantiations because we can't know which ones were intended to be overridden, and that seems... messy. So I guess we can accept this unusual approach for now, and see how it goes.

Reply:
So I think I felt it should be a hard error instead of a log because I'm imagining a typo or copy-paste error, and users being frustrated because it wasn't obvious that Terraform wasn't actually using the reattach mode, and so their plugin process wouldn't be getting attached to, but it would look like Terraform was working.

Failing in the command package later would be tricky, as when this fails, the only information we have is an opaque string we can't read out of; the JSON can't be read. So we'd need to change to parsing the JSON in the command package and passing the string in, and that would mean parsing the string more frequently, I think? My memory's foggy on that point.